sys/vfs.c

```diff
@@ -0,0 +1,226 @@
+}
+
+static int vfs_default_vget(mount_t *m, ino_t ino, vnode_t **v) {
+  return ENOTSUP;
+}
+
+static int vfs_default_init(vfsconf_t *vfc) {
+  return 0;
+}
+
+mount_t *vfs_mount_alloc(vnode_t *v, vfsconf_t *vfc) {
+  mount_t *m = kmalloc(vfs_pool, sizeof(mount_t), M_ZERO);
+
+  m->mnt_vfc = vfc;
+  m->mnt_vfsops = vfc->vfc_vfsops;
+  vfc->vfc_refcnt++; /* TODO: vfc_mtx? */
+  m->mnt_data = NULL;
+
+  m->mnt_vnodecovered = v;
+
+  m->mnt_refcnt = 0;
+  mtx_init(&m->mnt_mtx);
+
+  return m;
+}
+
+int vfs_domount(vfsconf_t *vfc, vnode_t *v) {
+  /* Start by checking whether this vnode can be used for mounting */
+  if (v->v_type != V_DIR)
+    return ENOTDIR;
+  if (v->v_mountedhere != NULL)
+    return EBUSY;
+
+  /* TODO: Mark the vnode is in-progress of mounting? See VI_MOUNT in FreeBSD */
+
+  mount_t *m = vfs_mount_alloc(v, vfc);
+
+  /* Mount the filesystem. */
+  int error = VFS_MOUNT(m);
+  if (error != 0) {
+    return error;
+  }
+
+  v->v_mountedhere = m;
+
+  mtx_lock(&mount_list_mtx);
+  TAILQ_INSERT_TAIL(&mount_list, m, mnt_list);
+  mtx_unlock(&mount_list_mtx);
+
+  /* Note that we do not need to ask the new mount for the root vnode! That
+     V_DIR vnode which is at the mount point stays in place. The root vnode is
+     requested during path lookup process, once we encounter a directory that
+     has v_mountedhere set. */
+
+  /* TODO: 'mountcheckdirs', which checks each process in the system, and
+     updates their dirs to reflect on the mounted file system... */
+
+  return 0;
+}
+
+int vfs_lookup(const char *path, vnode_t **vp) {
+
+  /* TODO: This is a simplified implementation, and it does not support many
+     required features! These include: relative paths, symlinks, parent dirs */
+
+  if (path[0] == '\0')
+    return EINVAL;
+
+  vnode_t *v;
+  if (strncmp(path, "/dev/", 5) == 0) {
+    /* Handle the special case of "/dev",
+     * since we don't have any filesystem at / (root) yet. */
+    v = vfs_root_dev_vnode;
+    path = path + 5;
+  } else if (strncmp(path, "/", 1) == 0) {
+    v = vfs_root_vnode;
+    path = path + 1;
+  } else {
+    log("Relative paths are not supported!");
+    return ENOTSUP;
+  }
+
+  /* Copy path into a local buffer, so that we may process it. */
+  size_t n = strlen(path);
+  if (n >= VFS_PATH_MAX)
+    return ENAMETOOLONG;
+  char pathcopy[VFS_PATH_MAX];
+  strlcpy(pathcopy, path, VFS_PATH_MAX);
+
+  vnode_lock(v);
+  vnode_hold(v);
+
+  const char *component;
+  char *pathbuf = pathcopy;
+  while ((component = strsep(&pathbuf, "/")) != NULL) {
+    if (component[0] == '\0')
+      continue;
+    /* If this vnode is a filesystem boundary,
+     * request the root vnode of the inner filesystem. */
+    if (v->v_mountedhere != NULL) {
+      mount_t *m = v->v_mountedhere;
+      vnode_t *v_mntpt;
+      int error = VFS_ROOT(m, &v_mntpt);
+      vnode_release(v);
+      vnode_unlock(v);
+      if (error != 0)
+        return error;
+      vnode_lock(v_mntpt);
+      v = v_mntpt;
```

Comment:
My original intention was that `VFS_ROOT` and `VOP_LOOKUP` should `hold` the vnode before returing it to the caller - since that's where the reference is copied (depending on implementation, a `VOP_LOOKUP` may return a reference without copying it, in that case it would not need to `hold` it again). For some reason, I forgot to implement that! Do you like this approach, or would you rather manually `vnode_hold(v_mntpt)` here?